### DISCUSSION TITLE
How to integrate wipay gateway in MERN Stack Project?

### DISCUSSION THREAD

**Discussion by GoldenDeveloper613** [community/community#147614] [Programming Help] [answered]
### Body

Hello, there.
Now I am developing e-commerce website using MERN Stack.
My Client requests to integrate wipay gateway to building website.
I received manual from him offered by Wipay Support team.
I searched through github repositories but couldn't find an exact answer.

If anyone integrated wipay gateway before, please help me. 

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by fdev0562** [comment, marked as answer]
Hello, GoldenDeveloper

I integrated wipay gateway before.
You can find that  here.
https://github.com/fdev0562/camm-backend/
https://github.com/fdev0562/camm-web/

This is main backend code for wipay.
//function to get redirect url to wipay payment page.
```
const fetch = require('node-fetch')
// const {Headers} = fetch
exports.getWipayRequestUrl = async (params) => {
  const body = new URLSearchParams();
  body.append('account_number', process.env.WIPAY_ACCOUNT_NUMBER);
  body.append('avs', '1');
  body.append('currency', "USD");
  body.append('environment', 'sandbox');
  body.append('origin', 'CammStock');
  body.append("country_code","BB")
  body.append('fee_structure', 'customer_pay');
  body.append('method', 'credit_card');
  body.append('data', JSON.stringify({a: "b"}));
  body.append('response_url', `${process.env.BACKEND_HOSTED_URL}/api/v1/purchase/completewipay`);
  for (let key in params) {
    body.append(key, params[key]);
  }

  // var headers = new Headers({
  //   'Accept': 'application/json'
  // });
  // headers.append('Accept', 'application/json');

  var options = {
    method: 'post',
    headers: {'Accept': 'application/json'},
    body: body,
    redirect: 'follow'
  };

  const rlt = await fetch('https://bb.wipayfinancial.com/plugins/payments/request', options);
  const text = await rlt.json();

  return text;
}
```

//controller function for payment and check payment status.

```
exports.createPurchaseReuestUrl = catchAsync(async (req, res, next) => {
  const { _id } = req.user;
  const {
    firstName,
    lastName,
    email,
    phoneNumber,
    country,
    state,
    city,
    zipCode,
    address,
  } = req.body;
  const wipayResponse = await getWipayRequestUrl({
    total: totalValue,
    fname: firstName,
    lname: lastName,
    addr1: address,
    email: email,
    phone: phoneNumber,
    country: getCountryCode(country),
    state: state,
    city: city,
    zipcode: zipCode,
    order_id: slug
  })
  if(wipayResponse?.url) {
    return res.status(200).json({url: wipayResponse.url});
  }
  return next(new AppError("Can't connect to payment gateway. Please try again later.", 400));
})

exports.completeWipayPayment = catchAsync(async (req, res, next) => {
  const {card, hash, customer_name, order_id, message, status, total, transaction_id} = req.query;
  if(status == "failed") {
    res.redirect(`${process.env.WEB_HOSTED_URL}/payment/checkout/cancel?message=${message}`);
    await Order.deleteOne({slug: order_id});
    return;
  }
  const orderList =await Order.findOne({slug: order_id});
  if(!orderList) {
    return res.redirect(`${process.env.WEB_HOSTED_URL}/payment/checkout/cancel?message=${'Something very went wrong.'}`);
  }
  const {amount, discount, purchaseData, user, coupon} = orderList;
  const calculatedHash = crypto.createHash('md5').update(`${transaction_id}${amount}${process.env.WIPAY_PRIVATE_KEY}`).digest('hex');
  if(calculatedHash != hash) {
    res.redirect(`${process.env.WEB_HOSTED_URL}/payment/checkout/cancel?message=${'Transaction not valid. Please contact Customer Support and Wipay with this key:' + order_id}`);
    return;
  }
  res.redirect(`${process.env.WEB_HOSTED_URL}/payment/checkout/success?message=${'Payment successful. Thank you for your purchase.'}`)
})
```

This is completly tested code.
You can update this code to integrate with your project.

**Comment by imrkaofficial** [comment]
1. Setup Backend (Node.js / Express)
You'll need to make API requests to Wipay’s server from your backend. For that, you'll likely interact with Wipay's API using HTTP requests (typically via axios or fetch).

Steps:

1. Install Dependencies: First, install the necessary packages to interact with Wipay's API and handle requests:

> `npm install axios`

2. Configure Wipay API Credentials: You’ll need your Wipay credentials (like API keys or secret keys) from the manual provided by Wipay support. Store them securely, ideally in environment variables.

3. Create an Endpoint to Handle Payment Requests: In your backend (Express server), create a route to handle the Wipay payment request.
`const express = require('express');
const axios = require('axios');
const dotenv = require('dotenv');

dotenv.config(); 

const app = express();
const PORT = process.env.PORT || 5000;

const WIPAY_API_URL = 'https://api.wipay.com/payment';
const API_KEY = process.env.WIPAY_API_KEY; 

app.use(express.json());

app.post('/api/payment', async (req, res) => {
    try {
        const paymentData = {
            amount: req.body.amount,
            currency: 'USD',
            order_id: req.body.orderId,
            customer_email: req.body.customerEmail,
        };

        const response = await axios.post(WIPAY_API_URL, paymentData, {
            headers: {
                'Authorization': `Bearer ${API_KEY}`,
                'Content-Type': 'application/json',
            },
        });

        res.json(response.data);
    } catch (error) {
        res.status(500).json({ error: 'Payment initiation failed', details: error.message });
    }
});

app.listen(PORT, () => {
    console.log(`Server is running on port ${PORT}`);
});`




2. Setup Frontend (React)
On the frontend, you’ll need to interact with this backend endpoint to initiate the payment process. Once the payment request is made, the frontend will redirect the user to Wipay’s hosted payment page.

Steps:

1. Install Axios (or another HTTP client for frontend): If you haven't installed axios already in your frontend project:

> `npm install axios`


2 . Create Payment Component: Create a component that will allow users to initiate the payment.

> `import React, { useState } from 'react';
import axios from 'axios';

const PaymentPage = () => {
    const [amount, setAmount] = useState('');
    const [orderId, setOrderId] = useState('');
    const [customerEmail, setCustomerEmail] = useState('');

    const handlePayment = async () => {
        try {
            const response = await axios.post('/api/payment', {
                amount: amount,
                orderId: orderId,
                customerEmail: customerEmail,
            });

            // Redirect user to Wipay payment page
            if (response.data.payment_url) {
                window.location.href = response.data.payment_url; // Redirect to Wipay
            }
        } catch (error) {
            console.error('Payment initiation failed:', error);
        }
    };

    return (
        <div>
            <h1>Make a Payment</h1>
            <input
                type="text"
                placeholder="Amount"
                value={amount}
                onChange={(e) => setAmount(e.target.value)}
            />
            <input
                type="text"
                placeholder="Order ID"
                value={orderId}
                onChange={(e) => setOrderId(e.target.value)}
            />
            <input
                type="email"
                placeholder="Customer Email"
                value={customerEmail}
                onChange={(e) => setCustomerEmail(e.target.value)}
            />
            <button onClick={handlePayment}>Pay Now</button>
        </div>
    );
};

export default PaymentPage;
`

3. Handle Wipay’s Response
Wipay will redirect the user back to your site after a payment attempt. Depending on the result (success, failure, etc.), you can handle this by capturing the response parameters (often in the query string) and updating your order status accordingly.

Create an endpoint to receive the callback (redirection) from Wipay.

> `app.get('/api/payment/callback', (req, res) => {
   const { status, transaction_id, order_id } = req.query;

   if (status === 'success') {
   } else {
   }

   res.redirect('/order-success'); // Redirect to a success page or failure page
});
`

4. Testing & Debugging
Thoroughly test the integration in a sandbox or staging environment provided by Wipay before going live.
Use tools like Postman to simulate the API requests and responses.

5. Security Considerations
Ensure that sensitive information (like your API keys) is stored securely, such as in environment variables.
Use HTTPS for all communication with the payment gateway to protect data.
Handle errors gracefully and securely.
